src/driver-steelseries.c

```diff
@@ -440,98 +508,84 @@ steelseries_write_led_v2(struct ratbag_led *led)
-		buf[27] = 0x01;
-		buf[28] = buf[31] = led->color.red;
-		buf[29] = buf[32] = led->color.green;
-		buf[30] = buf[33] = led->color.blue;
-
-		/* not sure why the duration is set for the steady color or why
-		   it is different for the two LEDs */
-		duration = led->index == 0 ? 10000 : 5000; /* 0x1027 or 0x8813 */
+		point1.color = led->color;
+
+		list_append(&cycle.points, &point1.link);
 		break;
 	case RATBAG_LED_CYCLE:
-		buf[27] = 0x04; /* number of steps in cycle */
-
-		/* start color */
-		buf[28] = buf[31] = 0xFF;
-		buf[29] = buf[32] = 0x00;
-		buf[30] = buf[33] = 0x00;
-
-		/* Cycle to green */
-		buf[35] = 0x00;
-		buf[36] = 0xFF;
-		buf[37] = 0x00;
-		buf[38] = 0x54; /* normalized time share of animation */
-
-		/* Cycle to blue */
-		buf[39] = 0x00;
-		buf[40] = 0x00;
-		buf[41] = 0xFF;
-		buf[42] = 0x54; /* normalized time share of animation */
-
-		/* Cycle to red */
-		buf[43] = 0xFF;
-		buf[44] = 0x00;
-		buf[45] = 0x00;
-		buf[46] = 0x56; /* normalized time share of animation */
-
-		duration = led->ms;
+		point1.color.red = 0xFF;
+		point1.color.green = 0x00;
+		point1.color.blue = 0x00;
+		point1.pos = 0x00;
+
+		point2.color.red = 0x00;
+		point2.color.green = 0xFF;
+		point2.color.blue = 0x00;
+		point2.pos = 0x54;
+
+		point3.color.red = 0x00;
+		point3.color.green = 0x00;
+		point3.color.blue = 0xFF;
+		point3.pos = 0x54;
+
+		point4.color.red = 0xFF;
+		point4.color.green = 0x00;
+		point4.color.blue = 0x00;
+		point4.pos = 0x56;
+
+		list_append(&cycle.points, &point1.link);
+		list_append(&cycle.points, &point2.link);
+		list_append(&cycle.points, &point3.link);
+		list_append(&cycle.points, &point4.link);
+
+		cycle.duration = led->ms;
 		break;
 	case RATBAG_LED_BREATHING:
-		buf[27] = 0x03; /* number of steps in cycle */
-
-		/* start color */
-		buf[28] = buf[31] = led->color.red;
-		buf[29] = buf[32] = led->color.green;
-		buf[30] = buf[33] = led->color.blue;
-
-		/* Cycle to black */
-		buf[35] = 0x00;
-		buf[36] = 0x00;
-		buf[37] = 0x00;
-		buf[38] = 0x7F; /* normalized time share of animation */
-
-		/* Cycle to selected color */
-		buf[39] = led->color.red;
-		buf[40] = led->color.green;
-		buf[41] = led->color.blue;
-		buf[42] = 0x7F; /* normalized time share of animation */
-
-		duration = led->ms;
+		point1.color.red = 0x00;
+		point1.color.green = 0x00;
+		point1.color.blue = 0x00;
+		point1.pos = 0x00;
+
+		point2.color = led->color;
+		point2.pos = 0x7F;
+
+		point3.color.red = 0x00;
+		point3.color.green = 0x00;
+		point3.color.blue = 0x00;
+		point3.pos = 0x7F;
+		
+		list_append(&cycle.points, &point1.link);
+		list_append(&cycle.points, &point2.link);
+		list_append(&cycle.points, &point3.link);
+
+		cycle.duration = led->ms;
 		break;
 	default:
 		return -EINVAL;
 	}
 
-	/* this seems to be the minimum allowed */
-	min_duration = buf[27] * 330;
-	if (duration < min_duration)
-		duration = min_duration;
-
-	hidpp_set_unaligned_le_u16(&buf[3], duration);
+	construct_cycle_buffer(&cycle, buf);
```

Comment:
for safety, always pass the buffer size with a buffer so you can easily assert in the function

Reply:
The buffer size is always the same. It's given by `STEELSERIES_REPORT_SIZE`. I can assert to that.

Reply:
pass the size in and your function becomes more flexible and less prone to erroneous copy/paste. Or even other changes when suddenly a profile buffer is different to a report buffer or whatever. The typing cost of passing the buffer size is close to zero, migth as well go for it :)